compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java

```diff
@@ -466,10 +470,91 @@ public BLangNode transform(TypeDefinitionNode typeDefNode) {
+                }
+
+                lastOfOthers.add(type);
+            }
+        }
+
+        List<TypeDescriptorNode> unionElements = new ArrayList<>();
+        reverseFlatMap(unionTypeElementsCollection, unionElements);
+
+        BLangFiniteTypeNode bLangFiniteTypeNode = (BLangFiniteTypeNode) TreeBuilder.createFiniteTypeNode();
+        for (TypeDescriptorNode finiteTypeEl : finiteTypeElements) {
+            SingletonTypeDescriptorNode singletonTypeNode = (SingletonTypeDescriptorNode) finiteTypeEl;
+            BLangLiteral literal = createSimpleLiteral(singletonTypeNode.simpleContExprNode());
+            bLangFiniteTypeNode.addValue(literal);
+        }
+
+        if (unionElements.isEmpty()) {
+            return bLangFiniteTypeNode;
+        }
 
-        return addUnionType(lhsTypeNode, rhsTypeNode, getPosition(unionTypeDescriptorNode));
+        BLangUnionTypeNode unionTypeNode = (BLangUnionTypeNode) TreeBuilder.createUnionTypeNode();
+        for (TypeDescriptorNode unionElement : unionElements) {
+            unionTypeNode.memberTypeNodes.add(createTypeNode(unionElement));
+        }
+
+        if (!finiteTypeElements.isEmpty()) {
+            unionTypeNode.memberTypeNodes.add(deSugarFiniteTypeAsUserDefType(bLangFiniteTypeNode));
+        }
+        return unionTypeNode;
+    }
+
+    private List<TypeDescriptorNode> flattenUnionType(UnionTypeDescriptorNode unionTypeDescriptorNode) {
+        List<TypeDescriptorNode> list = new ArrayList<>();
+        list.add(unionTypeDescriptorNode.leftTypeDesc());
+        while (unionTypeDescriptorNode.rightTypeDesc().kind() == SyntaxKind.UNION_TYPE_DESC) {
+            unionTypeDescriptorNode = (UnionTypeDescriptorNode) unionTypeDescriptorNode.rightTypeDesc();
+            list.add(unionTypeDescriptorNode.leftTypeDesc());
+        }
+        list.add(unionTypeDescriptorNode.rightTypeDesc());
+        return list;
+    }
+
+    private <T> void reverseFlatMap(List<List<T>> listOfLists, List<T> result) {
+        for (int i = listOfLists.size() - 1; i >= 0; i--) {
+            result.addAll(listOfLists.get(i));
+        }
+    }
+
+    private BLangUserDefinedType deSugarFiniteTypeAsUserDefType(BLangFiniteTypeNode toIndirect) {
+        DiagnosticPos pos = toIndirect.pos;
+        BLangUserDefinedType bLUserDefinedType;
+        BLangTypeDefinition bLTypeDef = (BLangTypeDefinition) TreeBuilder.createTypeDefinition();
+
+        // Generate a name for the anonymous object
+        String genName = anonymousModelHelper.getNextAnonymousTypeKey(diagnosticSource.pkgID);
+        IdentifierNode anonTypeGenName = createIdentifier(pos, genName);
+        bLTypeDef.setName(anonTypeGenName);
+        bLTypeDef.flagSet.add(Flag.PUBLIC);
+        bLTypeDef.flagSet.add(Flag.ANONYMOUS);
+
+        bLTypeDef.typeNode = toIndirect;
+        bLTypeDef.pos = pos;
+        currentCompilationUnit.addTopLevelNode(bLTypeDef);
```

Comment:
If you wan't to add another top-level node can achieve with `additionalTopLevelNodes.add(bLTypeDef)`, these nodes will be added to current compilation unit in [here](https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java#L340) or [here](https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java#L294).

Reply:
Ya I tried that, but the I got the types added in wrong oder. I wanted to get them in exact order as the old parser.

I'll improve this to keep the topNode, but have new private method `addToTop`, that will ignore it if it started form an inner node.

Reply:
How about adding them first? I noticed they are always on top of the list. How about mentioning the [index](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html#add(int,%20E)) with `additionalTopLevelNodes.add(<index>, bLTypeDef)` and maintain an index?

Reply:
I'll fix this with my next pr